pilopt/src/lib.rs

```diff
@@ -339,6 +340,23 @@ fn remove_trivial_identities<T: FieldElement>(pil_file: &mut Analyzed<T>) {
     pil_file.remove_identities(&to_remove);
 }
 
+fn remove_duplicate_identities<T: FieldElement>(pil_file: &mut Analyzed<T>) {
+    // Set of (left, right) tuples.
+    let mut identity_expressions = BTreeSet::new();
+    let to_remove = pil_file
+        .identities
+        .iter()
+        .enumerate()
+        .filter_map(|(index, identity)| {
```

Comment:
This could use `Itertools::duplicates_by`.

Reply:
Although I'm not sure about the documentation section "If an item appears more than twice, the second item is the item retained and the rest are discarded." - maybe add a test that has the same identity three times.

Reply:
Yes, it would only catch the second element: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=81b895f49a0da3c26c49a9c12ecea062

Reply:
I added a test though!